Replace kuka_plan_runner with iiwa_controller

sammy-tri opened the pull request:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/robotlocomotion/drake/5416)
<!-- Reviewable:end -->

sammy-tri commented:
+@siyuanfeng-tri for feature review.  

Note: the "do not merge" tag is here because I'm not sure what's going to happen once the system/dynamics people see this patch.  Heading off now to find a volunteer to take a look.

---

Review status: 0 of 6 files reviewed at latest revision, all discussions resolved.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/5416#-:-KeZcLINyZ5nIza0fJN8:b-a7t3e9)*
<!-- Sent from Reviewable.io -->

siyuanfeng-tri commented:
Reviewed 6 of 6 files at r1.
Review status: all files reviewed at latest revision, 4 unresolved discussions, some commit checks failed.

---

*[drake/examples/kuka_iiwa_arm/iiwa_controller.cc, line 42 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/5416#-KeZe7nqVaxDnGLApPK0:-KeZe7nqVaxDnGLApPK1:b-uyv96n) ([raw file](https://github.com/robotlocomotion/drake/blob/ef1080f3c576392c52196ec549ada5a7ff17b4c5/drake/examples/kuka_iiwa_arm/iiwa_controller.cc#L42)):*
> ```cpp
> const char* const kLcmPlanChannel = "COMMITTED_ROBOT_PLAN";
> const int kNumJoints = 7;
> ```

i would recommend add a block comment here say "i am doing a integrator on the interpolated reference position" or something similar. 

---

*[drake/examples/kuka_iiwa_arm/iiwa_controller.cc, line 44 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/5416#-KeZfwIoIpauVstbRKxt:-KeZfwIoIpauVstbRKxu:bgmhlds) ([raw file](https://github.com/robotlocomotion/drake/blob/ef1080f3c576392c52196ec549ada5a7ff17b4c5/drake/examples/kuka_iiwa_arm/iiwa_controller.cc#L44)):*
> ```cpp
> 
> int DoMain() {
>   lcm::DrakeLcm lcm;
> ```

BTW probably cleaner to make a diagram separately. but i am fine with this. 

---

*[drake/examples/kuka_iiwa_arm/iiwa_controller.cc, line 75 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/5416#-KeZfTvRkHFzIp3YhsL7:-KeZfTvRkHFzIp3YhsL8:b-ym47o) ([raw file](https://github.com/robotlocomotion/drake/blob/ef1080f3c576392c52196ec549ada5a7ff17b4c5/drake/examples/kuka_iiwa_arm/iiwa_controller.cc#L75)):*
> ```cpp
>   Eigen::VectorXd Kd = Eigen::VectorXd::Zero(kNumJoints);
> 
>   auto control_ports = systems::PidControlledSystem<double>::ConnectController(
> ```

i would recommend against systems::PidControlledSystem<double>::ConnectController, 
is there any reason why you don't want to do the connection part through the state_feedback_controller interface? 

---

*[drake/examples/kuka_iiwa_arm/iiwa_controller.cc, line 130 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/5416#-KeZeyqiRPVNjjIZWb8P:-KeZeyqiRPVNjjIZWb8Q:b-89akux) ([raw file](https://github.com/robotlocomotion/drake/blob/ef1080f3c576392c52196ec549ada5a7ff17b4c5/drake/examples/kuka_iiwa_arm/iiwa_controller.cc#L130)):*
> ```cpp
>   // status message.
>   while (true) {
>     if (next_control_time <= context->get_time()) {
> ```

BTW seems fine to me. @sherm1 should probably take a look at this chunk just to make sure we are not abusing anything. 

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/5416)*
<!-- Sent from Reviewable.io -->

patmarion commented:
nice!  has this been tested on hardware?  I would love to see some signal scope plots from this running on hardware with plans coming from director (starting from measured joint positions) :+1

sammy-tri commented:
It's basically the exact same code we were testing on the arm 2-3 weeks ago.  I've just been dragging my feet on the needed PRs (and doing a little refactoring) since then.

-(status: do not merge) 

---

Review status: 5 of 6 files reviewed at latest revision, 3 unresolved discussions.

---

*[drake/examples/kuka_iiwa_arm/iiwa_controller.cc, line 42 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/5416#-KeZe7nqVaxDnGLApPK0:-KeZoYsjCrbSdIRp31Hk:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/ef1080f3c576392c52196ec549ada5a7ff17b4c5/drake/examples/kuka_iiwa_arm/iiwa_controller.cc#L42)):*
<details><summary><i>Previously, siyuanfeng-tri wrote…</i></summary><blockquote>

i would recommend add a block comment here say "i am doing a integrator on the interpolated reference position" or something similar. 
</blockquote></details>

Done.

---

*[drake/examples/kuka_iiwa_arm/iiwa_controller.cc, line 75 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/5416#-KeZfTvRkHFzIp3YhsL7:-KeZmogBiTmR3vU-Ba-9:b-c7szpg) ([raw file](https://github.com/robotlocomotion/drake/blob/ef1080f3c576392c52196ec549ada5a7ff17b4c5/drake/examples/kuka_iiwa_arm/iiwa_controller.cc#L75)):*
<details><summary><i>Previously, siyuanfeng-tri wrote…</i></summary><blockquote>

i would recommend against systems::PidControlledSystem<double>::ConnectController, 
is there any reason why you don't want to do the connection part through the state_feedback_controller interface? 
</blockquote></details>

It wasn't there when I originally wrote this.  :) I do want the adder in pid_controlled_system.cc to add the output of the integrator to the feed-forward control input (which is the desired state in this case).  I can replicate that locally if we prefer that I use PidController directly.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/5416#-:-KeZmE6P60UetgrRjKca:bpev8nd)*
<!-- Sent from Reviewable.io -->

siyuanfeng-tri commented:
Review status: 5 of 6 files reviewed at latest revision, 2 unresolved discussions.

---

*[drake/examples/kuka_iiwa_arm/iiwa_controller.cc, line 75 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/5416#-KeZfTvRkHFzIp3YhsL7:-KeZqTJljw80JWl3Gbyy:b-vvare6) ([raw file](https://github.com/robotlocomotion/drake/blob/ef1080f3c576392c52196ec549ada5a7ff17b4c5/drake/examples/kuka_iiwa_arm/iiwa_controller.cc#L75)):*
<details><summary><i>Previously, sammy-tri (Sam Creasey) wrote…</i></summary><blockquote>

It wasn't there when I originally wrote this.  :) I do want the adder in pid_controlled_system.cc to add the output of the integrator to the feed-forward control input (which is the desired state in this case).  I can replicate that locally if we prefer that I use PidController directly.
</blockquote></details>

fine. this will become pretty confusing at some point... maybe add simple line comment here say like q_d = q_ref + ki * int (q_d - q). 

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/5416)*
<!-- Sent from Reviewable.io -->

siyuanfeng-tri commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

---







*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/5416#-:-KeZxfxvKa72l9hsfnZZ:bnfp4nl)*
<!-- Sent from Reviewable.io -->

amcastro-tri commented:
Review status: 5 of 6 files reviewed at latest revision, 3 unresolved discussions, some commit checks failed.

---

*[drake/examples/kuka_iiwa_arm/iiwa_controller.cc, line 131 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/5416#-Ke_tmBFbe1gQ6iUfxay:-Ke_tmBFbe1gQ6iUfxaz:btxiowp) ([raw file](https://github.com/robotlocomotion/drake/blob/8d2cb4ebd2f373391b222fa50b2331c7220c80f3/drake/examples/kuka_iiwa_arm/iiwa_controller.cc#L131)):*
> ```cpp
>   // Loop forever, continuing to update the time based on the incoming
>   // status message.
>   while (true) {
> ```

I think the way to accomplish this in a "system framework" manner, would be to create a something like a `SynchronizerSystem: public LeafSystem" that implements a `DoPublish()` and a `DoCalcNextUpdateTime()`. 
The `DoPublish()` method then would just wait (literally wait in real time) until time passes by unit `wall_clock_time > previous_time + dt` (maybe we need a state for `previous_time` in this system). Then the method `DoCalcNextUpdateTime()` would compute the next update time as `next_publish_time = wall_clock_time - context.ge_time()`.
Here I'm calling `wall_clock_time` the real time obtained from an LCM message. 
Then the way this would work is that you would wire-up your Diagram as you did above and add a "floating" (no inputs, no outputs) `SynchronizerSystem` to your Diagram. If I get it right I think that would have the same effect. I think @sherm1 actually did think about this scenario.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/5416)*
<!-- Sent from Reviewable.io -->

sherm1 commented:
+@edrumwri Evan, please take a look at iiwa_controller.cc to see if there is a way @sammy-tri can use the real Simulator to avoid having to attempt to duplicate its event handling code. I also added a few comments.

---

Review status: 5 of 6 files reviewed at latest revision, 6 unresolved discussions, some commit checks failed.

---

*[drake/examples/kuka_iiwa_arm/iiwa_controller.cc, line 121 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/5416#-KeaGXi3x13F6gBZQsHl:-KeaGXi3x13F6gBZQsHm:b-jfoxoa) ([raw file](https://github.com/robotlocomotion/drake/blob/8d2cb4ebd2f373391b222fa50b2331c7220c80f3/drake/examples/kuka_iiwa_arm/iiwa_controller.cc#L121)):*
> ```cpp
>   // integrator should do, precision is not critical for this
>   // application.
>   systems::ExplicitEulerIntegrator<double> integrator(
> ```

BTW, this is almost never the right integrator to use because of its terrible stability properties. Awful things will happen if the time step is large, which looks possible to me since it appears to be set from incoming message times. I would suggest switching to SemiExplicitEuler or RungeKutta2 to pick up a little more stability, and then set a maximum step size, taking multiple steps if necessary to get from one message time to the next. Or, let the Simulator take care of that for you.

---

*[drake/examples/kuka_iiwa_arm/iiwa_controller.cc, line 145 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/5416#-KeaFq7IzT20-ArQJBGK:-KeaFq7IzT20-ArQJBGL:b-4aktoh) ([raw file](https://github.com/robotlocomotion/drake/blob/8d2cb4ebd2f373391b222fa50b2331c7220c80f3/drake/examples/kuka_iiwa_arm/iiwa_controller.cc#L145)):*
> ```cpp
>           context->get_mutable_discrete_state()->CopyFrom(*discrete_state);
>         }
>       }
> ```

@edrumwri and I are concerned about the duplicated complexity of this "extract-of-Simulator" code. This messy code should appear in only one place. We are likely to change it in the Simulator and of course that will not be reflected here. I will add @edrumwri as a reviewer to see if there is a way this could use the actual Simulator to get this same job done.

---

*[drake/examples/kuka_iiwa_arm/iiwa_controller.cc, line 152 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/5416#-KeaEGYoD5Zt6lCrexVQ:-KeaEGYoD5Zt6lCrexVR:b-q7tpe4) ([raw file](https://github.com/robotlocomotion/drake/blob/8d2cb4ebd2f373391b222fa50b2331c7220c80f3/drake/examples/kuka_iiwa_arm/iiwa_controller.cc#L152)):*
> ```cpp
> 
>       const double dt = next_control_time - context->get_time();
>       integrator.StepOnceAtMost(dt, dt, dt);
> ```

BTW, the right method for this is `StepOnceExactly()` (please see doxygen). (Only for fixed-step integrators.)

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/5416#-:-KeaHOrVJkHay9fUdtOb:b-m70a5k)*
<!-- Sent from Reviewable.io -->

patmarion commented on drake/examples/kuka_iiwa_arm/iiwa_controller.cc:
```diff
@@ -0,0 +1,173 @@
+  // Loop forever, continuing to update the time based on the incoming
+  // status message.
+  while (true) {
+    if (next_control_time <= context->get_time()) {
```
a small suggestion, I think that for improved readability you could put the busy wait loop at the top, and then remove the `if` conditional so there isn't any nested code.

```
while (true) {

  while (busy wait condition) { }

   // do control

}
```

david-german-tri commented:
Review status: 5 of 6 files reviewed at latest revision, 7 unresolved discussions, some commit checks failed.

---

*[drake/examples/kuka_iiwa_arm/iiwa_controller.cc, line 131 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/5416#-Ke_tmBFbe1gQ6iUfxay:-KedCdt5t4gcNNO0ilJa:b6vkgxv) ([raw file](https://github.com/robotlocomotion/drake/blob/8d2cb4ebd2f373391b222fa50b2331c7220c80f3/drake/examples/kuka_iiwa_arm/iiwa_controller.cc#L131)):*
<details><summary><i>Previously, amcastro-tri (Alejandro Castro) wrote…</i></summary><blockquote>

I think the way to accomplish this in a "system framework" manner, would be to create a something like a `SynchronizerSystem: public LeafSystem" that implements a `DoPublish()` and a `DoCalcNextUpdateTime()`. 
The `DoPublish()` method then would just wait (literally wait in real time) until time passes by unit `wall_clock_time > previous_time + dt` (maybe we need a state for `previous_time` in this system). Then the method `DoCalcNextUpdateTime()` would compute the next update time as `next_publish_time = wall_clock_time - context.ge_time()`.
Here I'm calling `wall_clock_time` the real time obtained from an LCM message. 
Then the way this would work is that you would wire-up your Diagram as you did above and add a "floating" (no inputs, no outputs) `SynchronizerSystem` to your Diagram. If I get it right I think that would have the same effect. I think @sherm1 actually did think about this scenario.
</blockquote></details>

(a) As far as I can tell, this code has nothing to do with wall clock time.

(b) Please don't block or busy-wait in DoPublish as a way of gaming the Simulator loop.  That is not part of the System API's contract, and invites headaches when and if code other than Simulator needs to operate the System.  The right place to handle an external clock signal is the top-level loop.  Since that's not a feature Simulator has right now, it's reasonable to roll your own.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/5416)*
<!-- Sent from Reviewable.io -->

sammy-tri commented:
+(status: do not merge) +(status: do not review) Marking DNM and DNR since major structural changes are coming (to say the least).  Much thanks for the feedback which has been received so far, I'll address the existing issues before re-opening to review.

Feel free to leave additional comments while the DNR label is still on, it's more intended as an indicator that changes are coming.

---

Review status: 5 of 6 files reviewed at latest revision, 7 unresolved discussions, some commit checks failed.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/5416#-:-KedbDRYOJXV_84h-HCh:bu5j8nt)*
<!-- Sent from Reviewable.io -->

sammy-tri commented:
Updated with an alternate implementation which uses the simulator for small steps and tries to stay synced with the iiwa time.  It seems to work reasonably well with a faster-than-real-time simulation.

+(status: curate commits before merging)  -(status: do not review) -(status: do not merge) 

---

Review status: 3 of 6 files reviewed at latest revision, 6 unresolved discussions.

---

*[drake/examples/kuka_iiwa_arm/iiwa_controller.cc, line 121 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/5416#-KeaGXi3x13F6gBZQsHl:-Kep325tJ8Cmoggduzxx:b-trktes) ([raw file](https://github.com/robotlocomotion/drake/blob/8d2cb4ebd2f373391b222fa50b2331c7220c80f3/drake/examples/kuka_iiwa_arm/iiwa_controller.cc#L121)):*
<details><summary><i>Previously, sherm1 (Michael Sherman) wrote…</i></summary><blockquote>

BTW, this is almost never the right integrator to use because of its terrible stability properties. Awful things will happen if the time step is large, which looks possible to me since it appears to be set from incoming message times. I would suggest switching to SemiExplicitEuler or RungeKutta2 to pick up a little more stability, and then set a maximum step size, taking multiple steps if necessary to get from one message time to the next. Or, let the Simulator take care of that for you.
</blockquote></details>

Removed code.

---

*[drake/examples/kuka_iiwa_arm/iiwa_controller.cc, line 131 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/5416#-Ke_tmBFbe1gQ6iUfxay:-Kep2qqqml7g6JLppHn7:b-trktes) ([raw file](https://github.com/robotlocomotion/drake/blob/8d2cb4ebd2f373391b222fa50b2331c7220c80f3/drake/examples/kuka_iiwa_arm/iiwa_controller.cc#L131)):*
<details><summary><i>Previously, david-german-tri (David German) wrote…</i></summary><blockquote>

(a) As far as I can tell, this code has nothing to do with wall clock time.

(b) Please don't block or busy-wait in DoPublish as a way of gaming the Simulator loop.  That is not part of the System API's contract, and invites headaches when and if code other than Simulator needs to operate the System.  The right place to handle an external clock signal is the top-level loop.  Since that's not a feature Simulator has right now, it's reasonable to roll your own.
</blockquote></details>

Removed code.

---

*[drake/examples/kuka_iiwa_arm/iiwa_controller.cc, line 132 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/5416#-KeZc6_pKqTyuB4z3Be_-r2-132:-Kep2u2EVlu-lAsIpR5Q:bn53g51) ([raw file](https://github.com/robotlocomotion/drake/blob/8d2cb4ebd2f373391b222fa50b2331c7220c80f3/drake/examples/kuka_iiwa_arm/iiwa_controller.cc#L132)):*
<details><summary><i>Previously, patmarion (Pat Marion) wrote…</i></summary><blockquote>

a small suggestion, I think that for improved readability you could put the busy wait loop at the top, and then remove the `if` conditional so there isn't any nested code.

```
while (true) {

  while (busy wait condition) { }

   // do control

}
```
</blockquote></details>

The new version might be closer to what you were going for?

---

*[drake/examples/kuka_iiwa_arm/iiwa_controller.cc, line 145 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/5416#-KeaFq7IzT20-ArQJBGK:-Kep2y_6dLtquXGcYhY7:byge8md) ([raw file](https://github.com/robotlocomotion/drake/blob/8d2cb4ebd2f373391b222fa50b2331c7220c80f3/drake/examples/kuka_iiwa_arm/iiwa_controller.cc#L145)):*
<details><summary><i>Previously, sherm1 (Michael Sherman) wrote…</i></summary><blockquote>

@edrumwri and I are concerned about the duplicated complexity of this "extract-of-Simulator" code. This messy code should appear in only one place. We are likely to change it in the Simulator and of course that will not be reflected here. I will add @edrumwri as a reviewer to see if there is a way this could use the actual Simulator to get this same job done.
</blockquote></details>

Removed duplication.

---

*[drake/examples/kuka_iiwa_arm/iiwa_controller.cc, line 152 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/5416#-KeaEGYoD5Zt6lCrexVQ:-Kep3-AROUYE2q9GfJFV:b-trktes) ([raw file](https://github.com/robotlocomotion/drake/blob/8d2cb4ebd2f373391b222fa50b2331c7220c80f3/drake/examples/kuka_iiwa_arm/iiwa_controller.cc#L152)):*
<details><summary><i>Previously, sherm1 (Michael Sherman) wrote…</i></summary><blockquote>

BTW, the right method for this is `StepOnceExactly()` (please see doxygen). (Only for fixed-step integrators.)
</blockquote></details>

Removed code.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/5416#-:-Kep2V5mIL_rZExX1GkG:blr36uy)*
<!-- Sent from Reviewable.io -->

david-german-tri commented:
Review status: 3 of 6 files reviewed at latest revision, 4 unresolved discussions, some commit checks failed.

---

*[drake/examples/kuka_iiwa_arm/iiwa_controller.cc, line 131 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/5416#-Ke_tmBFbe1gQ6iUfxay:-KepAxOXFUn7j4GnIKhe:bcjmehw) ([raw file](https://github.com/robotlocomotion/drake/blob/8d2cb4ebd2f373391b222fa50b2331c7220c80f3/drake/examples/kuka_iiwa_arm/iiwa_controller.cc#L131)):*
<details><summary><i>Previously, sammy-tri (Sam Creasey) wrote…</i></summary><blockquote>

Removed code.
</blockquote></details>

This is a nice solution.  :+1:

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/5416)*
<!-- Sent from Reviewable.io -->

patmarion commented:
does StepTo always just take one step to catch up?  if not, the way i read this, if the controller took too long to run, it could fall behind further and further and stop publishing meaningful control signals? rather than just dropping a control tick here and there.  sorry if i misunderstand, i've not used the simulator api.

siyuanfeng-tri commented:
@patmarion good catch, i think you are correct... 

---







*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/5416#-:-KepRMdytqk5ufXEodo8:bv8nqxc)*
<!-- Sent from Reviewable.io -->

edrumwri commented:
Not sure what @patmarion means by "one step"- one integration step? If the simulator has to handle too many events in the time interval of dt, the simulation could fall behind. It's impossible to eliminate that possibility. This issue could be mitigated by treating all of the events as occurring at one point (e.g., the beginning or end of the interval), which would avoid numerous restarts.

---

Review status: 3 of 6 files reviewed at latest revision, 2 unresolved discussions, some commit checks failed.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/5416#-:-KepTpZM0YS5ezm8efAs:bmn9ejs)*
<!-- Sent from Reviewable.io -->

sherm1 commented:
Reviewed 3 of 6 files at r1, 2 of 3 files at r3, 1 of 1 files at r4.
Review status: all files reviewed at latest revision, 3 unresolved discussions, some commit checks failed.

---

*[drake/examples/kuka_iiwa_arm/BUILD, line 106 at r4](https://reviewable.io:443/reviews/robotlocomotion/drake/5416#-KepUwYZZqqpJmFKkFdH:-KepUwYZZqqpJmFKkFdI:boq27o1) ([raw file](https://github.com/robotlocomotion/drake/blob/de4260f13272ab30bdcdfa3d12134bada9997d17/drake/examples/kuka_iiwa_arm/BUILD#L106)):*
> ```Python
>         "//drake/common:text_logging_gflags",
>         "//drake/lcm",
>         "//drake/systems/analysis:explicit_euler_integrator",
> ```

BTW is this still a dependency?

---

*[drake/examples/kuka_iiwa_arm/iiwa_controller.cc, line 131 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/5416#-Ke_tmBFbe1gQ6iUfxay:-KepUTbUYyyyuRpV-0c1:bz0sfzz) ([raw file](https://github.com/robotlocomotion/drake/blob/8d2cb4ebd2f373391b222fa50b2331c7220c80f3/drake/examples/kuka_iiwa_arm/iiwa_controller.cc#L131)):*
<details><summary><i>Previously, david-german-tri (David German) wrote…</i></summary><blockquote>

This is a nice solution.  :+1:
</blockquote></details>

Looks great -- thanks, @sammy-tri!

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/5416)*
<!-- Sent from Reviewable.io -->

siyuanfeng-tri commented:
@patmarion On a second thought, I think we are ok if we set the controller's time step large enough. 

First of all, the controller is implemented as a discrete time system. In this particular "simulation", the only time consuming part is UnrestrictedStateUpdate, which is where we are doing control, and it is only run once every control_dt. The only state being integrated is the error on measured joint position vs some desired number, so CalcDerivative is trivial. 

The failure mode pat was referring to is where the controller take T (wall clock) > control_dt, in which case, the "simulation" runs slower than real time, and the controller lags more behind as time goes on. 

---







*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/5416#-:-KepVxdnN-zy_Vsd6kt6:bjqq198)*
<!-- Sent from Reviewable.io -->

patmarion commented:
`iiwa_controller.cc` doesn't set a dt, so what's the default behavior?  It looks like simulator.h has a default dt of 0.001, and `iiwa_controller.cc` sets the publish period of `command_pub` to `kIiwaLcmStatusPeriod=0.005`. 

When we call `simulator.StepTo(simulator.get_context().get_time() + 0.005)` how many times does the controller compute its output, and how many times does the `command_pub` publish an lcm message?   What if we call `StepTo(get_time() + 0.01)`, how many control updates and how many publishes then?

---

Review status: all files reviewed at latest revision, 3 unresolved discussions, some commit checks failed.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/5416#-:-KeqAi9D0wmIEv_WFCGo:b-4uvsxc)*
<!-- Sent from Reviewable.io -->

patmarion commented:
Review status: all files reviewed at latest revision, 4 unresolved discussions, some commit checks failed.

---

*[drake/examples/kuka_iiwa_arm/iiwa_controller.cc, line 126 at r4](https://reviewable.io:443/reviews/robotlocomotion/drake/5416#-KeqHAndKWYjQD--S-Ag:-KeqHAndKWYjQD--S-Ah:bs4g4kf) ([raw file](https://github.com/robotlocomotion/drake/blob/de4260f13272ab30bdcdfa3d12134bada9997d17/drake/examples/kuka_iiwa_arm/iiwa_controller.cc#L126)):*
> ```cpp
>       iiwa_time = static_cast<double>(msg->utime) / 1e6;
>     }
>   }
> ```

I still think you could change the order and remove the if condition, like:

```
while (True) {

  while (iiwa_time <= sim_time) {
      // busy wait
  }

  sim.StepTo(iiwa_time)

}
```

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/5416)*
<!-- Sent from Reviewable.io -->

siyuanfeng-tri commented:
@edrumwri or @sherm1 can you double check this? in case i am hallucinating. 

it's sort of complicated, a dt is set in https://github.com/RobotLocomotion/drake/blob/master/drake/examples/kuka_iiwa_arm/iiwa_plan_source.cc#L51. whenever ``context.get_time()`` hits ``kPlanUpdateInterval``, ``IiwaPlanSource::DoCalcUnrestrictedUpdate`` is executed, which is making a new pp if a new robot_plant_t has arrived. ``DoCalcOutput`` is a very lightweight interpolation. 

Now back to your question:
"When we call simulator.StepTo(simulator.get_context().get_time() + 0.005) how many times does the controller compute its output."
My answer is as many times as the integrator's eval time derivative is called. Because sammy is actually interpolating in ``DoCalcOutput``. But the heavy computation is only computed every ``kPlanUpdateInterval``. 
And if you move all computation into ``DoCalcUnrestrictedUpdate`` and have ``DoCalcOutput`` just copy from the abstract state, you can control exactly when everything is computed, and the output will be constant between updates.

"When we call simulator.StepTo(simulator.get_context().get_time() + dt) how many times does the controller compute its output, and how many times does the command_pub publish an lcm message?"
It depends on what integrator and what the integrator's max dt step is. If I only consider fixed step integrator, then ``DoCalcOutput`` is called by the number of the integrator's derivative calls within that dt window. And `` DoCalcUnrestrictedUpdate`` is called by the number of times context.get_time() crosses by ``kPlanUpdateInterval`` within that dt window. 
I think publish is actually actually called by default every time the integrator takes a step. So in this case, it will publish every 0.005, whether lcm can do that or not is a different question. 
 

---







*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/5416#-:-KeqHsnwGSW6IUUWojWF:b-mx9u3b)*
<!-- Sent from Reviewable.io -->

patmarion commented:
ok, thanks for the explanation!

I remain skeptical of using a simulation loop to wrap the iiwa control loop (at least, in the current form).

I don't want to block progress of this PR any further, but I request that the PR is updated so that `kuka_plan_runner` is not deleted.  I use `kuka_plan_runner` on the hardware and the behavior is well understood.  Of course, I am very eager to see `kuka_plan_runner` replaced with a controller, but the new controller needs to be proven on the hardware.  This PR also introduces the first usage of the IiwaPlanSource.  The way I read the code, it will cause the controller to produce step outputs that will cause the hardware to jump (potentially dangerously).

---

Review status: all files reviewed at latest revision, 4 unresolved discussions, some commit checks failed.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/5416#-:-KeshEcCI7jhxRcdxrMx:bgfwhup)*
<!-- Sent from Reviewable.io -->

edrumwri commented:
Siyuan's explanation is mostly correct: the number of loops will be equal to the number of "sample time hits" (due to publishes and updates) plus the number of integrator calls. The number of integrator calls can be limited by increasing the maximum step size (should be ok since the system looks amenable to integrating w/large steps), and the discussion above has already touched on minimizing publishes and updates.

---

Review status: all files reviewed at latest revision, 4 unresolved discussions, some commit checks failed.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/5416#-:-KesrprHp--cqVKF7w93:b-z93lco)*
<!-- Sent from Reviewable.io -->

sammy-tri commented:
I'll keep kuka_plan_runner around and as the default for now.  I've contacted @patmarion outside of this PR discussion to work out any issues with IiwaPlanSource.

-(status: curate commits before merging) 

---

Review status: 0 of 3 files reviewed at latest revision, 3 unresolved discussions.

---

*[drake/examples/kuka_iiwa_arm/BUILD, line 106 at r4](https://reviewable.io:443/reviews/robotlocomotion/drake/5416#-KepUwYZZqqpJmFKkFdH:-KetPu_1fjYhPmfHbC_I:bt2vr1b) ([raw file](https://github.com/robotlocomotion/drake/blob/de4260f13272ab30bdcdfa3d12134bada9997d17/drake/examples/kuka_iiwa_arm/BUILD#L106)):*
<details><summary><i>Previously, sherm1 (Michael Sherman) wrote…</i></summary><blockquote>

BTW is this still a dependency?
</blockquote></details>

Fixed.

---

*[drake/examples/kuka_iiwa_arm/iiwa_controller.cc, line 75 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/5416#-KeZfTvRkHFzIp3YhsL7:-KetQ9BgYGrqm_cbiuCj:bb6un1j) ([raw file](https://github.com/robotlocomotion/drake/blob/ef1080f3c576392c52196ec549ada5a7ff17b4c5/drake/examples/kuka_iiwa_arm/iiwa_controller.cc#L75)):*
<details><summary><i>Previously, siyuanfeng-tri wrote…</i></summary><blockquote>

fine. this will become pretty confusing at some point... maybe add simple line comment here say like q_d = q_ref + ki * int (q_d - q). 
</blockquote></details>

Done

---

*[drake/examples/kuka_iiwa_arm/iiwa_controller.cc, line 126 at r4](https://reviewable.io:443/reviews/robotlocomotion/drake/5416#-KeqHAndKWYjQD--S-Ag:-KetON5od-c6m3Y5MoAo:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/de4260f13272ab30bdcdfa3d12134bada9997d17/drake/examples/kuka_iiwa_arm/iiwa_controller.cc#L126)):*
<details><summary><i>Previously, patmarion (Pat Marion) wrote…</i></summary><blockquote>

I still think you could change the order and remove the if condition, like:

```
while (True) {

  while (iiwa_time <= sim_time) {
      // busy wait
  }

  sim.StepTo(iiwa_time)

}
```
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/5416#-:-KetOxESJo0MD1N9aPN4:bjeickl)*
<!-- Sent from Reviewable.io -->

manuelli commented:
I second Pat's comments. I really think that any changes to files that are run on hardware need to be extensively tested on the hardware before being merged. There is really no replacement for hardware testing.

sammy-tri commented:
It's been running on the hardware for a couple of hours so far today.  I'd advise against killing kuka_driver and leaving the controller + integrator running for an extended period of time, restarting the driver didn't do great things when I sent the first plan.  I could potentially add an abort case for significant jumps in iiwa_time (or an abort if the iiwa_time updating loop runs too many times without seeing the time change).

---

Review status: 0 of 3 files reviewed at latest revision, 3 unresolved discussions.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/5416#-:-KetvgJEUhUHGKEyf-xt:b9tfmz6)*
<!-- Sent from Reviewable.io -->

patmarion commented:
Review status: 0 of 3 files reviewed at latest revision, 3 unresolved discussions.

---

*[drake/examples/kuka_iiwa_arm/iiwa_controller.cc, line 111 at r5](https://reviewable.io:443/reviews/robotlocomotion/drake/5416#-KeuNNrVdMlDJEiRveV5:-KeuNNrWOd2JUEq-K9Ge:bbgczpk) ([raw file](https://github.com/robotlocomotion/drake/blob/3198d29e5ced29a3cdc6f7aca324bf1d4c102dcb/drake/examples/kuka_iiwa_arm/iiwa_controller.cc#L111)):*
> ```cpp
> 
>   systems::Simulator<double> simulator(*diagram, std::move(initial_context));
>   simulator.set_publish_every_time_step(false);
> ```

BTW, for those reviewing the code, the above line calling `set_publish_every_time_step(false)` was recently added after hardware tests today.  Without this line, the controller published 7 times for each `StepTo` call.

This code specifies a hardcoded publish period `kIiwaLcmStatusPeriod`.  I think that means if operate the kuka driver at a 1000hz rate instead of a 200hz rate then this controller will be 5x too slow.

I'd be interested to see code that implements a general LCM event loop to update a system exactly when new messages arrive, rather than using the simulator `StepTo` loop.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/5416)*
<!-- Sent from Reviewable.io -->

manuelli commented:
I want to second Pat's previous comment. This controller implementation is quite different from what we typically run on the robots. As Pat mentioned our standard setup is that the controller runs in an LCM event loop. The Atlas lcm event loop (threadedController.cpp) can be found [here](https://github.com/RobotLocomotion/valkyrie/blob/master/software/control/src/threadedController.cpp). This has worked really well for us for several years. 

The basic idea is that there is a set of LCM subscribers (e.g.  plan message, and the state estimator). Anytime a new message is received the controller is called using the newest of each message type and the control input is published out over LCM.

This is much simpler than the approach taken in the current PR. There is no simulator involved. I think if there was something that generalizes the LCM event loop in threadedController.cpp we would be really excited to use it. If there is interest Pat and I would be happy to have an in person meeting to explain how we use drake on the actual hardware. I think there are some important details that are different between hardware and simulation and maybe we can offer some unique insights on the hardware side since we both have the experience of the DRC under our belts.

sammy-tri commented:
I'd *definitely* prefer the ability to say something like "advance the time by whatever the delta is from the last time received via the LCM message we're using for the clock, and run *all* discrete/unrestricted updates whenever this happens".  I might be pushing to far on trying to share code between simulated drake robots and drake systems interacting with real robots...   my theory so far has been to make both cases share as much code as possible, but perhaps that should be something more like "as much code as is reasonable".

---

Review status: 0 of 3 files reviewed at latest revision, 3 unresolved discussions.

---

*[drake/examples/kuka_iiwa_arm/iiwa_controller.cc, line 111 at r5](https://reviewable.io:443/reviews/robotlocomotion/drake/5416#-KeuNNrVdMlDJEiRveV5:-Kf75mXLpePsJqFBDwi5:b-6rjf16) ([raw file](https://github.com/robotlocomotion/drake/blob/3198d29e5ced29a3cdc6f7aca324bf1d4c102dcb/drake/examples/kuka_iiwa_arm/iiwa_controller.cc#L111)):*
<details><summary><i>Previously, patmarion (Pat Marion) wrote…</i></summary><blockquote>

BTW, for those reviewing the code, the above line calling `set_publish_every_time_step(false)` was recently added after hardware tests today.  Without this line, the controller published 7 times for each `StepTo` call.

This code specifies a hardcoded publish period `kIiwaLcmStatusPeriod`.  I think that means if operate the kuka driver at a 1000hz rate instead of a 200hz rate then this controller will be 5x too slow.

I'd be interested to see code that implements a general LCM event loop to update a system exactly when new messages arrive, rather than using the simulator `StepTo` loop.



</blockquote></details>

kIiwaLcmStatus period needs to be adjusted to match the hardware, yes.  I responded a bit more in the discussion above.  Perhaps a comment indicating this adjustment would be appropriate.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/5416#-:-Kf74wpz762JfLYDXo5f:b-xy3sti)*
<!-- Sent from Reviewable.io -->

siyuanfeng-tri commented:
@sammy-tri there is nothing wrong with your approach, and that should definitely be preferred. we are not hitting anything fundamental yet about simulation vs hardware. 
the complaints from the robot people are mostly we'd like to have finer control / better understanding of what's going on, and less complicated a mechanism to advance time. reusing the Simulator api to do this is bit of an overkill, and there is too much hidden things happening. 

my vote for this particular pr is that it should go in to master (maybe dev), i think we have discussed this enough. but we need to figure out a more permanent solution. i am happy to start working on a lcm driven while true loop.  

---







*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/5416#-:-Kf7MAXmy-ZNtJ6YwBKB:b-ckiqii)*
<!-- Sent from Reviewable.io -->

patmarion commented:
yes i agree, this PR should be merged now.

---

Review status: 0 of 3 files reviewed at latest revision, 2 unresolved discussions.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/5416#-:-KfCd3tksGX4NQn0HjG9:b-jrss0b)*
<!-- Sent from Reviewable.io -->

sammy-tri commented:
@edrumwri any remaining issues?

---

Review status: 0 of 3 files reviewed at latest revision, all discussions resolved.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/5416#-:-KfCtWkgKm-fXTlEZRpj:b-xwdqe3)*
<!-- Sent from Reviewable.io -->

edrumwri commented:
No, <img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> 

---

Review status: 0 of 3 files reviewed at latest revision, all discussions resolved.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/5416#-:-KfCxoNI8JQr3IQ0W9Yq:benhfsq)*
<!-- Sent from Reviewable.io -->

sammy-tri commented:
+@david-german-tri for platform review (how did I forget that part?)

---

Review status: 0 of 3 files reviewed at latest revision, all discussions resolved.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/5416#-:-KfCyakIrHc5CnrhY4ZZ:bl21ptk)*
<!-- Sent from Reviewable.io -->

david-german-tri commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

---

Reviewed 3 of 6 files at r5.
Review status: all files reviewed at latest revision, all discussions resolved.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/5416#-:-KfDAOjbBiwY5ag5rIUr:bnfp4nl)*
<!-- Sent from Reviewable.io -->
